### PR TITLE
initial Metal support

### LINKED ISSUE #1304: Metal GPU acceleration on Apple Silicon
Hello,

I have a question.
Do you plan to support Metal GPU acceleration on Apple Silicon in future?

Thank you.

### PR DESCRIPTION
**Pull request by CarloLucibello**
Basic layers (Dense, Chain, ...) seem to be working fine. 
We don't have convolutions for the time being.

The Metal extension added here follows closely the structure of the AMDGPU one.

I'm adding a buildkite CI pipeline copied from https://github.com/JuliaGPU/Metal.jl/blob/main/.buildkite/pipeline.yml
(cc @maleadt)

Closes #1304

### CODE REVIEW

**Comment by CarloLucibello** [comment]
Metal tests are passing on buildkite. Tests have very limited coverage right now, but I think it ok for this initial PR. 
This is ready for review/merge.

**Comment by CarloLucibello** [comment]
buildkite v1 cuda failure is unrelated and probably a fluke

File: docs/src/gpu.md [hunk 1]
```diff
@@ -49,9 +60,12 @@ julia> Flux.GPU_BACKEND
 "CUDA"
 ```
 
+The current backend will affect the behaviour of methods like the method `gpu` described below.
+
 ## Basic GPU Usage
 
-Support for array operations on other hardware backends, like GPUs, is provided by external packages like [CUDA](https://github.com/JuliaGPU/CUDA.jl). Flux is agnostic to array types, so we simply need to move model weights and data to the GPU and Flux will handle it.
+Support for array operations on other hardware backends, like GPUs, is provided by external packages like [CUDA.jl](https://github.com/JuliaGPU/CUDA.jl), [AMDGPU.jl](https://github.com/JuliaGPU/AMDGPU.jl), and [Metal.jl](https://github.com/JuliaGPU/Metal.jl)
```
Comment (darsnack) [inline docs/src/gpu.md:67]:
```suggestion
Support for array operations on other hardware backends, like GPUs, is provided by external packages like [CUDA.jl](https://github.com/JuliaGPU/CUDA.jl), [AMDGPU.jl](https://github.com/JuliaGPU/AMDGPU.jl), and [Metal.jl](https://github.com/JuliaGPU/Metal.jl).
```

File: Project.toml [hunk 2]
```diff
@@ -58,7 +61,11 @@ Documenter = "e30172f5-a6a5-5a46-863b-614d45cd2de4"
...
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
+Metal = "dde4c033-4e86-420c-a63e-0dd931031962"
+Pkg = "44cfe95a-1eb2-52ea-b672-e2afdf69b78f"
 Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
 
 [targets]
-test = ["Test", "Documenter", "IterTools", "LinearAlgebra", "FillArrays", "ComponentArrays", "BSON"]
+test = ["Test", "Documenter", "IterTools", "LinearAlgebra", 
+        "FillArrays", "ComponentArrays", "BSON",
+        "Pkg"]
```
Comment (darsnack) [inline Project.toml:71]:
```suggestion
        "FillArrays", "ComponentArrays", "BSON", "Pkg"]
```

**Review by darsnack — APPROVED** [review]
Nice this is great!

**Comment by maleadt** [comment]
Feel free to file issues on Metal.jl for specific functionality that would be requires to unlock more of Flux.jl; it's good to know what to prioritize.

**Comment by ToucheSir** [comment]
I think the big one would be hooking up mpscnn support somehow, but that's not exactly specific and AFAIK would require significant effort.

**Pull request merged**